### PR TITLE
Problem: production rocksdb configuration is not optimal

### LINKED ISSUE #759: Problem: rocksdb compaction could cause write rate to drop to mimimum
**Is your feature request related to a problem? Please describe.**

rocksdb compaction could cause the write QPS to drop to minimum, which slows down the node performance.

**Describe the solution you'd like**

rocksdb 7.5.3 said that it fixed a bug related to `concurrent compactions might cause unnecessary further write stalling`, and this could possibly solve this problem.
maybe we could test and upgrade rocksdb to latest version to mitigate this.
https://github.com/facebook/rocksdb/releases/tag/v7.5.3

> Bug Fixes
>      * Fix a bug in which backup/checkpoint can include a WAL deleted by RocksDB.
>      * `Fix a bug where concurrent compactions might cause unnecessary further write stalling. In some cases, this might cause write rate to drop to minimum.`

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- update related dependencies to allow custom rocksdb options.
- using rocksdb v7, closes #759 
- tune rocksdb options.

During the versiondb migration experiments, I found the default rocksdb settings are multiple times slower than the one I experimented, because I've tuned options and did a full compaction on the db before, but after I keep the node running with the normal binary for a while, suddenly my script's performance drop significantly, and the sst files are all rewritten by compaction process.
I suspect the main difference is sst file sizes, and probably filter and compression types as well, so I plan to tune the rocksdb options for default binary first, then try the versiondb migration script again, to see if it can get back to the old speed.

### Compare Compression Options

- Tool: `sst_dump --file=sample.sst --command=recompress ...`
- Block size: 32k

| Compression Type                    | Size      | Time  |
| ----------------------------------- | --------- | ----- |
| Snappy  | 228861267 | 1.71s |
| LZ4                                 | 228341872 | 1.85s |
| ZSTD                                | 217266835 | 3.93  |
| ZSTD level 11 | 214701135 | 13.7s |
| ZSTD level 11 with 11k dictionary | 186643202 | 20.6s |
| ZSTD level 11 with 110k dictionary   | 177305609 | 22.4s |

Basically I picked relatively slower speed and higher compression ratio for the bottommost level, I think our node do have the capacity to occupy one core to do the heavier compaction gradually.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,6 @@ NETWORK ?= mainnet
 LEDGER_ENABLED ?= true
 # RocksDB is a native dependency, so we don't assume the library is installed.
 # Instead, it must be explicitly enabled and we warn when it is not.
-ENABLE_ROCKSDB ?= false
```
Comment (yihuang) [inline Makefile:11]: This is option is actually for the store v2, useless in foreseeable future.

File: Makefile [hunk 2]
```diff
@@ -8,7 +8,6 @@ NETWORK ?= mainnet
 LEDGER_ENABLED ?= true
 # RocksDB is a native dependency, so we don't assume the library is installed.
 # Instead, it must be explicitly enabled and we warn when it is not.
```
Comment (yihuang) [inline Makefile:10]:
```suggestion
```

File: cmd/cronosd/cmd/open_db_rocksdb.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+	// compression options at bottommost level
+	opts.SetBottommostCompression(grocksdb.ZSTDCompression)
+	compressOpts := grocksdb.NewDefaultCompressionOptions()
+	compressOpts.MaxDictBytes = 112640 // 110k
+	compressOpts.Level = 12
+	opts.SetBottommostCompressionOptions(compressOpts, true)
+	opts.SetBottommostCompressionOptionsZstdMaxTrainBytes(compressOpts.MaxDictBytes*100, true)
+	return opts
+}
```
[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/cronosd/cmd/open_db_rocksdb.go:58]: I forgot to include `IncreaseParallelism` and `OptimizeLevelStyleCompaction`, done, those are kept the same with existing settings.

[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/cronosd/cmd/open_db_rocksdb.go:58]: > No SetMaxOpenFiles, allowing DB backend open files as much as possible.

file descriptors should not be a problem, if ram usage is a concern we should enable `SetCacheIndexAndFilterBlocks`, and with bigger sst file size (`SetTargetFileSizeMultiplier`) and block size (32k), the total amount of filter and index should be reduced.

[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/cronosd/cmd/open_db_rocksdb.go:58]: > How do we decide the MaxDictBytes and level for ZSTD?

I added some compression comparison in PR description, just based on empirical data.
I read somewhere the train bytes can be 100 * dict bytes, and no need to set dict bytes larger than 110k, so I picked the biggest number in the recommended range.

[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/cronosd/cmd/open_db_rocksdb.go:58]: > SetTargetFileSizeMultiplier , the default is 1. How do we decide whether this number is 2 or not if we want to reduce the total file number?

the base level is 64M (set by `OptimizeLevelStyleCompaction(512m)`), so I think multiply with 2 for each level is big enough, it goes like: `128M`, `256M`, `512M`...

**Review by mmsqe — COMMENTED** [review]
still get `could not determine kind of name for C.rocksdb_lru_cache_options_set_num_shard_bits` build error in grocksdb

**Comment by yihuang** [comment]
> still get `could not determine kind of name for C.rocksdb_lru_cache_options_set_num_shard_bits` build error in grocksdb

try recent version of rocksdb, or using nix, the nix script use `v7.8.3`.

**Review by mmsqe — APPROVED** [review]

**Comment by yihuang** [comment]
@JayT106 I'll merge now to unblock some other fixes, feel free to further discuss the options picked here.

**Comment by JayT106** [comment]
>file descriptors should not be a problem, if ram usage is a concern we should enable SetCacheIndexAndFilterBlocks, and with bigger sst file size (SetTargetFileSizeMultiplier) and block size (32k), the total amount of filter and index should be reduced.

I remembered the value `SetMaxOpenFiles` was affecting the ram usage a lot when exporting the genesis file. >64GB using when the value is `4096` and <32GB using when it is `1000`
Maybe we need to evaluate it against the new settings. Before the genesis export feature is settled in the SDK, it will always be an issue.

**Comment by JayT106** [comment]
> the base level is 64M (set by OptimizeLevelStyleCompaction(512m)), multiply with 2 for each level seems big enough, it goes like: 128M, 256M, 512M...
I guess there are other downsides with too big sst file, like slow down compaction, or bigger filter/index, i'm not 100% sure actually.
Maybe we should try bigger value since mainnet is much bigger than testnet.

My understanding based on your setup with `linxGnu/grocksdb` will be:
```
                           L1     L2     L3     L4     L5     L6     L7 ...
PerFileSize:              1MB    2MB    4MB    8MB   16MB   32MB   64MB
MaxDataBytes:            10MB  100MB    1GB   10GB  100GB    1TB   10TB
EstMaxFile#:               10     50    250   1250   6250  31250 156250
```

**Comment by yihuang** [comment]
> > file descriptors should not be a problem, if ram usage is a concern we should enable SetCacheIndexAndFilterBlocks, and with bigger sst file size (SetTargetFileSizeMultiplier) and block size (32k), the total amount of filter and index should be reduced.
> 
> I remembered the value `SetMaxOpenFiles` was affecting the RAM usage a lot when exporting the genesis file. >64GB using when the value is `4096` and <32GB using when it is `1000` Maybe we need to evaluate it against the new settings. Before the genesis export feature is settled in the SDK, it will always be an issue.

can you try `SetCacheIndexAndFilterBlocks(true)` to control ram usage during genesis export?
I think the reason open file limit affect ram usage is because the index/filter caches.

**Comment by JayT106** [comment]
> > > file descriptors should not be a problem, if ram usage is a concern we should enable SetCacheIndexAndFilterBlocks, and with bigger sst file size (SetTargetFileSizeMultiplier) and block size (32k), the total amount of filter and index should be reduced.
> > 
> > 
> > I remembered the value `SetMaxOpenFiles` was affecting the RAM usage a lot when exporting the genesis file. >64GB using when the value is `4096` and <32GB using when it is `1000` Maybe we need to evaluate it against the new settings. Before the genesis export feature is settled in the SDK, it will always be an issue.
> 
> can you try `SetCacheIndexAndFilterBlocks(true)` to control ram usage during genesis export? I think the reason open file limit affect ram usage is because the index/filter caches, `CacheIndexAndFilterBlocks` will manage those caches together with block cache.

tried cronosd`1.0.2-3-gca5e564` on the dev machine has 64GB RAM Recent datasets `cronosmainnet_25-1-rocksdb-pruned.20230122.2040.tar.lz4`, OOM.   Will try the same code with limited `SetMaxOpenFiles`

**Comment by yihuang** [comment]
> > > > file descriptors should not be a problem, if ram usage is a concern we should enable SetCacheIndexAndFilterBlocks, and with bigger sst file size (SetTargetFileSizeMultiplier) and block size (32k), the total amount of filter and index should be reduced.
> 
> > > 
> 
> > > 
> 
> > > I remembered the value `SetMaxOpenFiles` was affecting the RAM usage a lot when exporting the genesis file. >64GB using when the value is `4096` and <32GB using when it is `1000` Maybe we need to evaluate it against the new settings. Before the genesis export feature is settled in the SDK, it will always be an issue.
> 
> > 
> 
> > can you try `SetCacheIndexAndFilterBlocks(true)` to control ram usage during genesis export? I think the reason open file limit affect ram usage is because the index/filter caches, `CacheIndexAndFilterBlocks` will manage those caches together with block cache.
> 
> 
> 
> tried cronosd`1.0.2-3-gca5e564` on the dev machine has 64GB RAM Recent datasets `cronosmainnet_25-1-rocksdb-pruned.20230122.2040.tar.lz4`, OOM.   Will try the same code with limited `SetMaxOpenFiles` 
> 
> 
> 
> 
> 
> 

Can you try `SetCacheIndexAndFilterBlocks(true)`?

**Comment by yihuang** [comment]
> > > > file descriptors should not be a problem, if ram usage is a concern we should enable SetCacheIndexAndFilterBlocks, and with bigger sst file size (SetTargetFileSizeMultiplier) and block size (32k), the total amount of filter and index should be reduced.
> > > 
> > > 
> > > I remembered the value `SetMaxOpenFiles` was affecting the RAM usage a lot when exporting the genesis file. >64GB using when the value is `4096` and <32GB using when it is `1000` Maybe we need to evaluate it against the new settings. Before the genesis export feature is settled in the SDK, it will always be an issue.
> > 
> > 
> > can you try `SetCacheIndexAndFilterBlocks(true)` to control ram usage during genesis export? I think the reason open file limit affect ram usage is because the index/filter caches, `CacheIndexAndFilterBlocks` will manage those caches together with block cache.
> 
> tried cronosd`1.0.2-3-gca5e564` on the dev machine has 64GB RAM Recent datasets `cronosmainnet_25-1-rocksdb-pruned.20230122.2040.tar.lz4`, OOM. Will try the same code with limited `SetMaxOpenFiles`

Is genesis export OOM or node syncing OOM? @borischeuk-crypto has tried on one archive node, didn't see much change in runtime statistics.

**Pull request merged**